### PR TITLE
Calculate HH progress C++ port

### PR DESCRIPTION
**Pull request by RZR-UA**

### CODE REVIEW

**Comment by RZR-UA** [comment]
https://github.com/osmandapp/OsmAnd/issues/18969

File: native/src/routeCalculationProgress.h [hunk 1]
```diff
@@ -70,25 +71,52 @@ class RouteCalculationProgress {
...
+		hhIterationStep = step;
+		hhCurrentStepProgress = 0;
+	}
+
+	virtual void hhTargetsProgress(int done, int total) {
+		hhTargetsDone = done;
+		hhTargetsTotal = total;
+	}
+
+	virtual void hhIterationProgress(double k) {
```
Comment (ivanPyrohivskyi) [inline native/src/routeCalculationProgress.h:114]: Only declaration here

File: native/src/routePlannerFrontEnd.h [hunk 2]
```diff
@@ -110,7 +110,7 @@ class RoutePlannerFrontEnd {
 	bool pointCloseEnough(SHARED_PTR<GpxRouteApproximation>& gctx, SHARED_PTR<GpxPoint>& ipoint,
 	                      vector<SHARED_PTR<RouteSegmentResult>>& res);
 	void makeSegmentPointPrecise(SHARED_PTR<RouteSegmentResult>& routeSegmentResult, double lat, double lon, bool st);
-    HHNetworkRouteRes * calculateHHRoute(HHRoutePlanner & routePlanner, int startX, int startY, int endX, int endY, double dir);
+    HHNetworkRouteRes * calculateHHRoute(HHRoutePlanner & routePlanner, RoutingContext* ctx, int startX, int startY, int endX, int endY, double dir);
```
Comment (ivanPyrohivskyi) [inline native/src/routePlannerFrontEnd.h:113]: Formatting

File: native/src/routeCalculationProgress.cpp [hunk 3]
```diff
@@ -97,7 +98,34 @@ void RouteCalculationProgress::updateStatus(float distanceFromBegin, int directS
...
+			} else {
+				progress += (float) found->second; // passed step
+			}
+		}
+	}
+
+	if (hhTargetsTotal > 0) {
+		progress = (100.0 * (float) hhTargetsDone + progress) / (float) hhTargetsTotal; // intermediate points
+	} else {
+		OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Error, "Error: getLinearProgressHH() got hhTargetsTotal=0");
```
Comment (ivanPyrohivskyi) [inline native/src/routeCalculationProgress.cpp:119]: Check cancel route and remove if redunant

File: native/src/routeCalculationProgress.h (hunk 1 above)
Comment (RZR-UA) [reply native/src/routeCalculationProgress.h:114]: functions moved to cpp

File: native/src/routeCalculationProgress.h [hunk 4]
```diff
@@ -70,25 +71,52 @@ class RouteCalculationProgress {
...
+		SELECT_REGIONS,
+		LOAD_POINTS,
+		START_END_POINT,
+		ROUTING,
+		DETAILED,
+		ALTERNATIVES,
+		DONE
+	};
+
+	const std::unordered_map<HHIteration, int> hhIterationPercent = {
```
[the remark this answers is not in the document]
Comment (RZR-UA) [reply native/src/routeCalculationProgress.h:89]: refactored to switch/case, kept in .h file near the enum

File: native/src/routeCalculationProgress.cpp (hunk 3 above)
Comment (RZR-UA) [reply native/src/routeCalculationProgress.cpp:119]: this logger should be kept until release version

File: native/src/routePlannerFrontEnd.cpp [hunk 5]
```diff
@@ -802,7 +802,8 @@ vector<SHARED_PTR<RouteSegmentResult>> RoutePlannerFrontEnd::searchRoute(
         HHNetworkRouteRes * r = nullptr;
         double dir = ctx->config->initialDirection ;
         for (int i = 0; i < targetsX.size(); i++) {
-            res = calculateHHRoute(routePlanner, i == 0 ? startX : targetsX.at(i - 1),
```
[the remark this answers is not in the document]
Comment (RZR-UA) [reply native/src/routePlannerFrontEnd.cpp:805]: fixed

File: native/src/routePlannerFrontEnd.h (hunk 2 above)
Comment (RZR-UA) [reply native/src/routePlannerFrontEnd.h:113]: fixed

File: native/src/routeCalculationProgress.cpp [hunk 6]
```diff
@@ -97,7 +98,34 @@ void RouteCalculationProgress::updateStatus(float distanceFromBegin, int directS
 	this->reverseSegmentQueueSize = reverseSegmentQueueSize;
 }
 
+float RouteCalculationProgress::getLinearProgressHH() {
+	float progress = 0;
+
+	for (int i = HHIteration::HH_NOT_STARTED; i <= HHIteration::DONE; i++) {
```
Comment (ivanPyrohivskyi) [inline native/src/routeCalculationProgress.cpp:104]: According to best practices on C++ better to add to HHIteration "Count" item

File: native/src/routePlannerFrontEnd.h [hunk 7]
```diff
@@ -42,8 +42,8 @@ class RoutePlannerFrontEnd {
     
     bool useSmartRouteRecalculation;
     
-    bool TRACE_ROUTING = false;
-    bool USE_HH_ROUTING = false;
+//    bool TRACE_ROUTING = false; // not used
```
Comment (ivanPyrohivskyi) [inline native/src/routePlannerFrontEnd.h:45]: Just remove

Comment (RZR-UA) [reply native/src/routePlannerFrontEnd.h:45]: done

File: native/src/routeCalculationProgress.cpp (hunk 6 above)
Comment (RZR-UA) [reply native/src/routeCalculationProgress.cpp]: Count added

**Pull request merged**